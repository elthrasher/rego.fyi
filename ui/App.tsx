import { createMuiTheme, Link, ThemeProvider, Typography } from '@material-ui/core';
import Container from '@material-ui/core/Container';
import Grid from '@material-ui/core/Grid';
import React from 'react';

import Rego from './pages/Rego';
import { PayloadsProvider } from './providers/PayloadsProvider';

const theme = createMuiTheme({
  typography: {
    fontSize: 10,
  },
});

const App = (): JSX.Element => {
  return (
    <PayloadsProvider>
      <Container>
        <Grid container alignItems="center">
          <Grid item lg={2} sm={12}>
            <Container>
              <Typography>
                <ThemeProvider theme={theme}>
                  <h4>What is this?</h4>
                  <p>
                    I was interested to see if I could use rego in a serverless environment. I put together a demo that
                    involved using a REST client, copying tokens around and stuff like that. I decided my demo was
                    terrible and resolved to do something better.
                  </p>
                  <h4>How do I use it?</h4>
                  <p>
                    Edit any of the three text blocks to the right and select an HTTP method and path. The top and
                    bottom should have JSON data and are validated. The middle one is written in{' '}
                    <Link
                      href="https://www.openpolicyagent.org/docs/latest/policy-language/"
                      rel="noopener"
                      target="_blank"
                    >
                      rego
                    </Link>{' '}
                    and isn't validated (yet) but you could use{' '}
                    <Link href="https://play.openpolicyagent.org/" rel="noopener" target="_blank">
                      The Rego Playground
                    </Link>
                    .
                  </p>
                  <p>
                    When you click the SEND REQUEST button, the app will encode all three text blocks and make an HTTP
                    request to the serverless backend. The backend will decode the token, compile the rego policy,
                    evaluate the policy with the other payloads, then return the appropriate response.
                  </p>
                  <h4>Production use?</h4>
                  <p>
                    This? Nope. You don't want your authorization system to accept a policy from the client. Something
                    like this? Very possibly.
                  </p>
                </ThemeProvider>
              </Typography>
            </Container>
          </Grid>
          <Grid item lg={8} sm={12}>
            <Rego />
          </Grid>
          <Grid item lg={2} sm={12}>
            <Container>
              <Typography>
                <ThemeProvider theme={theme}>
                  <h4>Architecture</h4>
                  <p>
                    The web app is written in React and is an S3 website. The API uses API Gateway and Lambda. Nothing
                    sent to the API is stored, but there's some logging in place.
                  </p>
                  <p>
                    The backend consists of two Lambda functions, an authorizer written in golang and a handler written
                    in NodeJS. The authorizer is implemented with{' '}
                    <Link href="https://github.com/open-policy-agent/opa" rel="noopener" target="_blank">
                      opa packages
                    </Link>
                    .
                  </p>
                  <h4>Open Policy Agent</h4>
                  <p>
                    I'm not affiliated.{' '}
                    <Link href="https://www.openpolicyagent.org/" rel="noopener" target="_blank">
                      Learn more.
                    </Link>
                  </p>
                </ThemeProvider>
              </Typography>
            </Container>
          </Grid>
        </Grid>
      </Container>
    </PayloadsProvider>
  );
};

export default App;
