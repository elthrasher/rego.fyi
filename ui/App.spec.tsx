import React from 'react';
import renderer from 'react-test-renderer';

import App from './App';

it('renders a message', () => {
  const component = renderer.create(<App />, {
    createNodeMock: (node) => {
      return document.createElement(node.type.toString());
    },
  });

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
