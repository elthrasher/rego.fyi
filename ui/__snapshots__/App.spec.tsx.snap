// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a message 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthLg"
>
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
      >
        <h1>
          rego.fyi
        </h1>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-9"
      />
    </div>
  </div>
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
      >
        <div
          className="MuiFormControl-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
            data-shrink={true}
          >
            Method
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <div
              aria-haspopup="listbox"
              className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiInputBase-input MuiInput-input"
              onBlur={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              role="button"
              tabIndex={0}
            >
              GET
            </div>
            <input
              aria-hidden={true}
              className="MuiSelect-nativeInput"
              onAnimationStart={[Function]}
              onChange={[Function]}
              required={false}
              tabIndex={-1}
              value="GET"
            />
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSelect-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6"
      >
        <div
          className="MuiFormControl-root"
        >
          <div
            className="MuiFormControl-root MuiTextField-root"
          >
            <label
              className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
              data-shrink={true}
            >
              Resource
            </label>
            <div
              className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
              onClick={[Function]}
            >
              <input
                aria-invalid={false}
                autoFocus={false}
                className="MuiInputBase-input MuiInput-input"
                disabled={false}
                onAnimationStart={[Function]}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                required={false}
                type="text"
                value="/hello"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <fieldset
      className="MuiFormControl-root MuiFormControl-fullWidth"
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled"
          data-shrink={true}
          htmlFor="User info such as a JWT or session"
          id="User info such as a JWT or session-label"
        >
          User info such as a JWT or session
        </label>
        <div
          className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiFilledInput-multiline"
          onClick={[Function]}
        >
          <textarea
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputMultiline MuiFilledInput-inputMultiline"
            disabled={false}
            id="User info such as a JWT or session"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            rows={1}
            style={
              Object {
                "height": undefined,
                "overflow": null,
              }
            }
            value="{\\"role_name\\":\\"Admin\\"}"
          />
          <textarea
            aria-hidden={true}
            className="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputMultiline MuiFilledInput-inputMultiline"
            readOnly={true}
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "hidden",
                "position": "absolute",
                "top": 0,
                "transform": "translateZ(0)",
                "visibility": "hidden",
              }
            }
            tabIndex={-1}
          />
        </div>
      </div>
    </fieldset>
  </div>
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <fieldset
      className="MuiFormControl-root MuiFormControl-fullWidth"
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled"
          data-shrink={true}
          htmlFor="Rego policy loaded from disk or other storage"
          id="Rego policy loaded from disk or other storage-label"
        >
          Rego policy loaded from disk or other storage
        </label>
        <div
          className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiFilledInput-multiline"
          onClick={[Function]}
        >
          <textarea
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputMultiline MuiFilledInput-inputMultiline"
            disabled={false}
            id="Rego policy loaded from disk or other storage"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            rows={1}
            style={
              Object {
                "height": undefined,
                "overflow": null,
              }
            }
            value="package rolemapper

import data.roles

default allow = false

allow {
    role_has_permission[input.role_name]
}

role_has_permission[role_name] {
    r = roles[_]
    r.name = role_name
    match_with_wildcard(r.operations, input.operation)
    match_with_wildcard(r.resources, input.resource)
}
match_with_wildcard(allowed, value) {
    allowed[_] = \\"*\\"
}
match_with_wildcard(allowed, value) {
    allowed[_] = value
}"
          />
          <textarea
            aria-hidden={true}
            className="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputMultiline MuiFilledInput-inputMultiline"
            readOnly={true}
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "hidden",
                "position": "absolute",
                "top": 0,
                "transform": "translateZ(0)",
                "visibility": "hidden",
              }
            }
            tabIndex={-1}
          />
        </div>
      </div>
    </fieldset>
  </div>
  <div
    className="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <fieldset
      className="MuiFormControl-root MuiFormControl-fullWidth"
    >
      <div
        className="MuiFormControl-root MuiTextField-root MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-filled"
          data-shrink={true}
          htmlFor="Data loaded from disk or other storage for making policy decisions"
          id="Data loaded from disk or other storage for making policy decisions-label"
        >
          Data loaded from disk or other storage for making policy decisions
        </label>
        <div
          className="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-multiline MuiFilledInput-multiline"
          onClick={[Function]}
        >
          <textarea
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputMultiline MuiFilledInput-inputMultiline"
            disabled={false}
            id="Data loaded from disk or other storage for making policy decisions"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            rows={1}
            style={
              Object {
                "height": undefined,
                "overflow": null,
              }
            }
            value="{\\"roles\\":[{\\"resources\\":[\\"/hello\\"],\\"operations\\":[\\"GET\\"],\\"name\\":\\"Admin\\"}]}"
          />
          <textarea
            aria-hidden={true}
            className="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputMultiline MuiFilledInput-inputMultiline"
            readOnly={true}
            style={
              Object {
                "height": 0,
                "left": 0,
                "overflow": "hidden",
                "position": "absolute",
                "top": 0,
                "transform": "translateZ(0)",
                "visibility": "hidden",
              }
            }
            tabIndex={-1}
          />
        </div>
      </div>
    </fieldset>
  </div>
  <button
    className="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-label"
    >
      Send Request
    </span>
  </button>
</div>
`;
